src/main.cpp

```
@@ -0,0 +1,1605 @@
+
+    queue_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+    queue_create_info.queueFamilyIndex = family_indices.drawing_family;
+    queue_create_info.queueCount = 1;
+    queue_create_info.pQueuePriorities = &queue_priority;
+    queue_create_info_vector.push_back(queue_create_info);
+
+    queue_create_info = {};
+    queue_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+    queue_create_info.queueFamilyIndex = family_indices.presentation_family;
+    queue_create_info.queueCount = 1;
+    queue_create_info.pQueuePriorities = &queue_priority;
+    queue_create_info_vector.push_back(queue_create_info);
+
+    device_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+    device_create_info.pQueueCreateInfos = queue_create_info_vector.data();
+    device_create_info.queueCreateInfoCount = queue_create_info_vector.size();
+    device_create_info.pEnabledFeatures = &device_features;
+    device_create_info.ppEnabledExtensionNames = required_extensions.data();
+    device_create_info.enabledExtensionCount = required_extensions.size();
+    device_create_info.enabledLayerCount = 0;
+
+    result = vkCreateDevice(
+            device,
+            &device_create_info,
+            nullptr,
+            &logical_device);
+    if (VK_SUCCESS != result) {
+        throw std::runtime_error("Failed to create logical device");
+    }
+
+    return logical_device;
+}
+
+static VkSurfaceFormatKHR
+pick_swap_surface_format(
+        const std::vector<VkSurfaceFormatKHR>& available_formats)
+{
+    bool should_pick;
+
+    should_pick = false;
+
+    for (const VkSurfaceFormatKHR& format : available_formats) {
+        should_pick =
+            VK_FORMAT_B8G8R8A8_SRGB == format.format &&
+            VK_COLOR_SPACE_SRGB_NONLINEAR_KHR == format.colorSpace;
+        if (should_pick) {
+            return format;
+        }
+    }
+
+    Log::w("Could not find the preferred format combination");
+
+    return available_formats[0];
```

Please throw, instead of returning.

---

Modified in: https://github.com/pottumuusi/neotetris/pull/1/commits/3ecd3c17c1d78aacbf1bc468defce974a0df56bc